flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/TaskMailboxImpl.java

```diff
@@ -77,22 +112,30 @@ public boolean hasMail() {
 
 	@Override
 	public Optional<Mail> tryTake(int priority) throws IllegalStateException {
+		Optional<Mail> head = tryTakeFromBatch();
```

Comment:
shouldn't the `tryTake` and `take` methods call `createBatch()` if `tryTakeFromBatch` returned `Optional.empty()`?

Why do we allow for non batched takes? Shouldn't we have just `tryTakeFromBatch` and `takeFromBatch`?

Reply:
Ok, but in that case:

- please add a comment about this contract somewhere in this class
- and either please add an ITCase coverage for the case when two operators are ping ponging themselves with adding mails and yielding, which will test that `tryTake` & `take` do not expand the batch.
- or add additional comment inside the existing unit test, why is it testing for that contract

Reply:
Clarified the contract in `TaskMailbox`. And added a test based on the stream task test harness that checks that only mails in the current batch are taken. Also added more comment to the unit tests that tests the interaction on a fine-grain level.